test/src/main/java/org/apache/accumulo/test/fate/accumulo/FateStoreIT.java

```diff
@@ -216,7 +227,104 @@ protected void testDeferredOverflow(FateStore<TestEnv> store, ServerContext sctx
+    }
+  }
+
+  @Test
+  public void testCreateWithKey() throws Exception {
+    executeTest(this::testCreateWithKey);
+  }
+
+  protected void testCreateWithKey(FateStore<TestEnv> store, ServerContext sctx) {
+    KeyExtent ke1 = new KeyExtent(TableId.of("tableId"), new Text("zzz"), new Text("aaa"));
+    KeyExtent ke2 = new KeyExtent(TableId.of("tableId2"), new Text("zzz"), new Text("aaa"));
+
+    byte[] key1 = serialize(ke1);
+    long tid1 = store.create(key1);
+
+    byte[] key2 = serialize(ke2);
+    long tid2 = store.create(key2);
+    assertNotEquals(tid1, tid2);
+
+    FateTxStore<TestEnv> txStore1 = store.reserve(tid1);
+    FateTxStore<TestEnv> txStore2 = store.reserve(tid2);
+    try {
+      assertTrue(txStore1.timeCreated() > 0);
+      assertEquals(TStatus.NEW, txStore1.getStatus());
+      assertArrayEquals(key1, txStore1.getKey().orElseThrow());
+
+      assertTrue(txStore2.timeCreated() > 0);
+      assertEquals(TStatus.NEW, txStore2.getStatus());
+      assertArrayEquals(key2, txStore2.getKey().orElseThrow());
+
+      assertEquals(2, store.list().count());
+    } finally {
+      txStore1.delete();
+      txStore2.delete();
+    }
+  }
+
+  @Test
+  public void testCreateWithKeyDuplicate() throws Exception {
+    executeTest(this::testCreateWithKeyDuplicate);
+  }
+
+  protected void testCreateWithKeyDuplicate(FateStore<TestEnv> store, ServerContext sctx) {
+    KeyExtent ke = new KeyExtent(TableId.of("tableId"), new Text("zzz"), new Text("aaa"));
+
+    // Creating with the same key should be fine if the status is NEW
+    // It should just return the same id and allow us to continue reserving
+    byte[] key = serialize(ke);
+    long tid1 = store.create(key);
+    long tid2 = store.create(key);
+    assertEquals(tid1, tid2);
+
+    FateTxStore<TestEnv> txStore = store.reserve(tid1);
+    try {
+      assertTrue(txStore.timeCreated() > 0);
+      assertEquals(TStatus.NEW, txStore.getStatus());
+      assertArrayEquals(key, txStore.getKey().orElseThrow());
+      assertEquals(1, store.list().count());
+    } finally {
+      txStore.delete();
+    }
+  }
+
+  @Test
+  public void testCreateWithKeyInProgress() throws Exception {
+    executeTest(this::testCreateWithKeyInProgress);
+  }
+
+  protected void testCreateWithKeyInProgress(FateStore<TestEnv> store, ServerContext sctx) {
+    KeyExtent ke = new KeyExtent(TableId.of("tableId"), new Text("zzz"), new Text("aaa"));
+
+    byte[] key = serialize(ke);
+    long tid1 = store.create(key);
+
+    FateTxStore<TestEnv> txStore = store.reserve(tid1);
+    try {
+      assertTrue(txStore.timeCreated() > 0);
+      txStore.setStatus(TStatus.IN_PROGRESS);
+
+      // We have an existing transaction with the same key in progress
+      // so should not be allowed
+      assertThrows(IllegalStateException.class, () -> store.create(key));
+    } finally {
+      txStore.delete();
+    }
+  }
+
```

Comment:
Would be good to test the case of having a collision of the id when the keys differ.  Thinking we could do the following in test code so that no test specific modifications are needed to real code.

 1. Compute H1 = hash(K1) for key K1.  Use the same hash algo as store impls.
 2. In test code manually insert K2 where K1!=K2 in the data layer of the store using id H1 from above.
 3. Try to insert K1 using the store, should fail because K2 exist w/ same id.

Not sure about code organization for step 2 because it will need to be different test code depending on if Zookeeper or Accumulo store is being tested.

Reply:
In order to make this easier to test I updated the store to allow providing a different hash code or FateId generated from the key and i replace it for the test to see if the collision detection worked